File: torch-c-article/lib/th/TestModule.lua
require('torch')
require('nn')

--- @class TestModule
local TestModule, TestModuleParent = torch.class('nn.TestModule', 'nn.Module')

--- Initialize.
function TestModule:__init(threshold)
  self.threshold = threshold or 0.5
end

--- Compute forward pass, i.e. threshold to 1 at 0.1.
-- @param input layer input
-- @param output
function TestModule:updateOutput(input)
  assert(input:type() == 'torch.FloatTensor')
  assert(lib.cpu)

  self.output = input:clone()
  lib.cpu.test_module_updateOutput(5, input:size():data(), input:data(), self.output:data())
  return self.output
end

--- Compute the backward pass.
-- @param input original input
-- @param gradOutput gradients of top layer
-- @return gradients with respect to input
function TestModule:updateGradInput(input, gradOutput)
  assert(input:type() == 'torch.FloatTensor')
  assert(lib.cpu)

  self.gradInput = gradOutput:clone()
  lib.cpu.test_module_updateGradInput(5, input:size():data(), input:data(), gradOutput:data(), self.output:data())
  return self.gradInput
end

File: torch-c-article/lib/cpp/cpu/test_module.cpp
#include <cstdio>
#include <cassert>
#include "test_module.h"

void test_module_updateOutput(const int rank, const long* dims, const float* input, float* output) {
  assert(rank == 5);

  for (int r = 0; r < rank; r++) {
    printf("%ld ", dims[r]);
  }
  printf("\n");

  const int batch_size = dims[0];
  const int channels = dims[1];
  const int height = dims[2];
  const int width = dims[3];
  const int depth = dims[4];

  for (int b = 0; b < batch_size; b++) {
    printf("batch %d\n", b);
    for (int c = 0; c < channels; c++) {
      printf("channel %d\n", c);
      for (int h = 0; h < height; h++) {
        printf("height %d\n", h);
        for (int w = 0; w < width; w++) {
          for (int d = 0; d < depth; d++) {
            printf("%f ", input[(((b*channels + c)*height + h)*width + w)*depth + d]);
          }
          printf("\n");
        }
      }
    }
  }
}

void test_module_updateGradInput(const int rank, const long* dims, const float* input, const float* grad_output, float* grad_input) {
  assert(rank == 5);

  for (int r = 0; r < rank; r++) {
    printf("%ld ", dims[r]);
  }
  printf("\n");

  const int batch_size = dims[0];
  const int channels = dims[1];
  const int height = dims[2];
  const int width = dims[3];
  const int depth = dims[4];

  for (int b = 0; b < batch_size; b++) {
    printf("batch %d\n", b);
    for (int c = 0; c < channels; c++) {
      printf("channel %d\n", c);
      for (int h = 0; h < height; h++) {
        printf("height %d\n", h);
        for (int w = 0; w < width; w++) {
          for (int d = 0; d < depth; d++) {
            printf("%f ", input[(((b*channels + c)*height + h)*width + w)*depth + d]);
          }
          printf("\n");
        }
      }
    }
  }
}